include/CL/sycl/parallelism/detail/parallelism.hpp

```diff
@@ -239,35 +239,47 @@ void parallel_for_workitem(const group<Dimensions> &g,
 
      The issue is that the parallel_for_workitem() execution is slow even
      when nd_item::barrier() is not used
-
-     \todo Simplify by just using omp parallel for collapse
   */
-
+  T_Item index{g.get_nd_range()};
   range<Dimensions> l_r = g.get_nd_range().get_local_range();
-  auto tot = l_r.get(0);
-  for (int i = 1; i < (int) Dimensions; ++i) {
-    tot *= l_r.get(i);
-  }
-#pragma omp parallel num_threads(tot)
-  {
-    T_Item index { g.get_nd_range() };
-    id<Dimensions> local; // to initialize correctly
-#pragma omp for nowait
-    for (std::size_t th_id = 0; th_id < tot; ++th_id) {
-      if (Dimensions == 1) {
-        local[0] = th_id;
-      } else if (Dimensions == 2) {
-        local[0] = th_id / l_r.get(1);
-        local[1] = th_id % l_r.get(1);
-      } else if (Dimensions == 3) {
-        local[0] = th_id / (l_r.get(1)*l_r.get(2));
-        local[1] = (th_id / l_r.get(2)) % l_r.get(1);
-        local[2] = th_id % l_r.get(2);
-      }
-      index.set_local(local);
-      index.set_global(local + id<Dimensions>(l_r)*g.get_id());
+  id<Dimensions> temp, id_l_r = id<Dimensions>(l_r);
+
+  if constexpr (Dimensions == 1) {
+#pragma omp parallel for collapse(1) firstprivate(index, temp)
+    for (size_t i = 0; i < l_r.get(0); ++i) {
+      temp[0] = i;
+      index.set_local(temp);
+      temp[0] = i + id_l_r[0] * g.get_id()[0];
+      index.set_global(temp);
       f(index);
     }
+  } else if constexpr (Dimensions == 2) {
+#pragma omp parallel for collapse(1) firstprivate(index, temp)
+    for (size_t i = 0; i < l_r.get(0); ++i) {
+      #pragma omp simd collapse(1)
+      for (size_t j = 0; j < l_r.get(1); ++j) {
+        temp[0] = i; temp[1] = j;
+        index.set_local(temp);
+        temp[0] = i + id_l_r[0] * g.get_id()[0];
+        temp[1] = j + id_l_r[1] * g.get_id()[1];
+        index.set_global(temp);
+        f(index);
+      }
+    }
+  } else if constexpr (Dimensions == 3) {
+    #pragma omp parallel for collapse(2) firstprivate(index, temp)
+    for (size_t i = 0; i < l_r.get(0); ++i)
+      for (size_t j = 0; j < l_r.get(1); ++j)
+      #pragma omp simd collapse(1)
+        for (size_t k = 0; k < l_r.get(2); ++k) {
+          temp[0] = i; temp[1] = j; temp[2] = k;
+          index.set_local(temp);
```

Comment:
`index.set_local({ i, j, k });` ?

Reply:
Or '{ k, j, i }' ? I have a doubt...

Reply:
I believe it's i, j, k as index[2] should increment the quickest, index[1] the second quickest and index[0] the slowest.

At least based off of the previous code, assuming a square 4x4x4 work-group:
```
local[0] = th_id / (l_r.get(1)*l_r.get(2));
local[1] = (th_id / l_r.get(2)) % l_r.get(1);
local[2] = th_id % l_r.get(2);  
```
local[0] should increment every 16 threads: 16 / (4*4) = 1, anything that's a decimal should be rounded down by casting I think
local[1] should increment every 4 threads: (4 / 4) % 4 = 1
local[2] should increment every iteration and reset to 0 at 4 as it's just the remainder of the division of 4 into the thread id, so 4 % 4 = 0 and 3 % 4 = 3

That's my understanding of it anyway, it took me a little time to work it out as it's not so simple at a glance!